expression/collation.go

```diff
@@ -189,12 +168,103 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
+	case ast.InsertFunc:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[0], args[3])
+	case ast.Lpad, ast.Rpad:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[0], args[2])
+	case ast.Elt, ast.ExportSet, ast.MakeSet:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[1:]...)
+	case ast.FindInSet:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args...)
+	case ast.Field:
+		if argTps[0] == types.ETString {
+			return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args...)
+		}
+	case ast.Locate, ast.Instr:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[0], args[1])
+	case ast.GE, ast.LE, ast.GT, ast.LT, ast.EQ, ast.NE, ast.NullEQ:
+		// if compare type is string, we should determine which collation should be used.
+		if argTps[0] == types.ETString {
+			dstCharset, dstCollation, _, err = CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args...)
+			return dstCharset, dstCollation, CoercibilityNumeric, err
+		}
+	case ast.If:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[1], args[2])
+	case ast.Like:
+		dstCharset, dstCollation, _, err = CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args[0], args[1])
+		if err != nil {
+			return
+		}
+		return dstCharset, dstCollation, CoercibilityCoercible, err
+	case ast.Regexp:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args...)
+	case ast.In:
+		if args[0].GetType().EvalType() == types.ETString {
+			return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args...)
+		}
+	case ast.DateFormat, ast.TimeFormat:
+		charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
+		return charsetInfo, collation, args[1].Coercibility(), nil
+	case ast.Cast:
+		// we assume all the cast are implicit.
+		if retType == types.ETString {
+			charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
+			return charsetInfo, collation, args[0].Coercibility(), nil
+		}
+		return charset.CharsetBin, charset.CollationBin, args[0].Coercibility(), nil
+	case ast.Case:
+		// FIXME: case function aggregate collation is not correct.
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args...)
+	case ast.Database, ast.User, ast.CurrentUser, ast.Version, ast.CurrentRole, ast.TiDBVersion:
+		chs, coll := charset.GetDefaultCharsetAndCollate()
+		return chs, coll, CoercibilitySysconst, nil
+	}
+
+	if retType == types.ETString {
+		charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
+		return charsetInfo, collation, CoercibilityCoercible, nil
+	}
+	return charset.CharsetBin, charset.CollationBin, CoercibilityNumeric, nil
+}
+
 // DeriveCollationFromExprs derives collation information from these expressions.
+// Deprecated, use CheckAndDeriveCollationFromExprs instead.
```

Comment:
Is this function planned for obsolescence? Roughly what can be removed? If so, do you need to add a TODO?

Reply:
sure, I will add a TODO, I will remove this function in the GBK developing